cpp/src/io/parquet/experimental/page_index_filter_utils.cu

```diff
@@ -0,0 +1,253 @@
+                                        ? offset_index.page_locations[page_idx + 1].first_row_index
+                                        : row_group.num_rows;
+                                    auto const page_size = last_row_idx - first_row_idx;
+                                    max_page_size = std::max<size_type>(max_page_size, page_size);
+                                    page_row_offsets.push_back(page_row_offsets.back() + page_size);
+                                  });
+                  });
+                });
+
+  return {std::move(page_row_offsets), max_page_size};
+}
+
+rmm::device_uvector<size_type> compute_page_indices_async(
+  cudf::host_span<cudf::size_type const> page_row_counts,
+  cudf::host_span<cudf::size_type const> page_row_offsets,
+  cudf::size_type total_rows,
+  rmm::cuda_stream_view stream)
+{
+  auto mr = cudf::get_current_device_resource_ref();
+
+  // Copy page-level row counts and offsets to device
+  auto row_counts  = cudf::detail::make_device_uvector_async(page_row_counts, stream, mr);
+  auto row_offsets = cudf::detail::make_device_uvector_async(page_row_offsets, stream, mr);
+
+  // Make a zeroed device vector to store page indices of each row
+  auto page_indices =
+    cudf::detail::make_zeroed_device_uvector_async<cudf::size_type>(total_rows, stream, mr);
+
+  // Scatter page indices across the their first row's index
+  thrust::scatter_if(rmm::exec_policy_nosync(stream),
+                     thrust::counting_iterator<size_type>(0),
+                     thrust::counting_iterator<size_type>(row_counts.size()),
+                     row_offsets.begin(),
+                     row_counts.begin(),
+                     page_indices.begin());
+
+  // Inclusive scan with maximum to replace zeros with the (increasing) page index it belongs to.
+  // Page indices are scattered at their first row's index.
+  thrust::inclusive_scan(rmm::exec_policy_nosync(stream),
+                         page_indices.begin(),
+                         page_indices.end(),
+                         page_indices.begin(),
+                         cuda::maximum<cudf::size_type>());
+  return page_indices;
+}
+
+std::vector<size_type> compute_fenwick_tree_level_offsets(cudf::size_type level0_size,
+                                                          cudf::size_type max_page_size)
+{
+  std::vector<size_type> tree_level_offsets;
+  tree_level_offsets.push_back(0);
+
+  cudf::size_type current_level_size = cudf::util::div_rounding_up_unsafe(level0_size, 2);
+  cudf::size_type current_level      = 1;
+
+  while (current_level_size > 0) {
+    auto const block_size = 1 << current_level;
+    if (std::cmp_greater(block_size, max_page_size)) { break; }
+    tree_level_offsets.push_back(tree_level_offsets.back() + current_level_size);
+    current_level_size =
+      current_level_size == 1 ? 0 : cudf::util::div_rounding_up_unsafe(current_level_size, 2);
```

Comment:
`cudf::util::div_rounding_up_unsafe` will not let the current_level_size to become zero so manually checking